substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1059 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Extension module for Sassafras consensus.
+//!
+//! Sassafras is a constant-time block production protocol that aims to ensure that
+//! there is exactly one block produced with constant time intervals rather multiple
+//! or none.
+//!
+//! We run a lottery to distribute block production slots in an epoch and to fix the
+//! order validators produce blocks by the beginning of an epoch.
+//!
+//! Each validator signs the same VRF input and publish the output onchain. This
+//! value is their lottery ticket that can be validated against their public key.
+//!
+//! We want to keep lottery winners secret, i.e. do not publish their public keys.
+//! At the begin of the epoch all the validators tickets are published but not their
+//! public keys.
+//!
+//! A valid tickets are validated when an honest validator reclaims it on block
+//! production.
+//!
+//! To prevent submission of fake tickets, resulting in empty slots, the validator
+//! when submitting the ticket accompanies it with a SNARK of the statement: "Here's
+//! my VRF output that has been generated using the given VRF input and my secret
+//! key. I'm not telling you my keys, but my public key is among those of the
+//! nominated validators", that is validated before the lottery.
+//!
+//! To anonymously publish the ticket to the chain a validator sends their tickets
+//! to a random validator who later puts it on-chain as a transaction.
+
+#![deny(warnings)]
+#![warn(unused_must_use, unsafe_code, unused_variables, unused_imports, missing_docs)]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use log::{debug, error, warn};
+use scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+
+use frame_support::{
+	dispatch::{DispatchResultWithPostInfo, Pays},
+	traits::Get,
+	weights::Weight,
+	BoundedVec, WeakBoundedVec,
+};
+use frame_system::{
+	offchain::{SendTransactionTypes, SubmitTransaction},
+	pallet_prelude::BlockNumberFor,
+};
+use sp_consensus_sassafras::{
+	digests::{ConsensusLog, NextEpochDescriptor, SlotClaim},
+	vrf, AuthorityId, Epoch, EpochConfiguration, Randomness, Slot, TicketBody, TicketEnvelope,
+	TicketId, RANDOMNESS_LENGTH, SASSAFRAS_ENGINE_ID,
+};
+use sp_io::hashing;
+use sp_runtime::{generic::DigestItem, traits::One, BoundToRuntimeAppPublic};
+use sp_std::prelude::Vec;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(all(feature = "std", test))]
+mod mock;
+#[cfg(all(feature = "std", test))]
+mod tests;
+
+pub mod weights;
+pub use weights::WeightInfo;
+
+pub use pallet::*;
+
+const LOG_TARGET: &str = "sassafras::runtime 🌳";
+
+// Contextual string used by the VRF to generate per-block randomness.
+const RANDOMNESS_VRF_CONTEXT: &[u8] = b"SassafrasRandomness";
```

Comment:
This is only used in the pallet to update the on-chain randomness according to its private rules (which are encapsulated here)

I don't see any usage for this in other components (e.g. for the client) at the moment.
For sure if it will be useful somewhere I'll move this to the primitives

Reply:
I changed the value to "SassafrasOnchainRandomness"... 
Just in case the client needs to build some other (unrelated) randomness from the same VrfOutput source